util/check_dif_statuses.py

```diff
@@ -0,0 +1,314 @@
+            num_funcs_implemented = colored(
+                dif_status.num_functions_implemented, "red")
+
+        # Add row to table (printing one contributor email per line).
+        rows.append([
+            dif_status.ip, dif_last_modified, hw_last_modified,
+            "\n".join(dif_status.dif_main_contributors),
+            "\n".join(dif_status.hw_main_contributors), num_funcs_defined,
+            num_funcs_implemented, dif_status.lifecycle_state
+        ])
+
+    # Print the table and legend.
+    print("DIF Statuses:")
+    print(tabulate(rows, headers, tablefmt=TABLE_FORMAT))
+    print("""*Only the top two contributors (by LOC) """
+          """for each component are listed.""")
+    print(colored("Yellow", "yellow"),
+          "\t= HW has been updated since the DIF.")
+    print(
+        colored("Green", "green"),
+        """\t= DIF API, as defined in the current header file, is complete. """
+        """Note, the header file may lack necessary API functionality.""")
+    print(colored("Red", "red"),
+          "\t= DIF API is incomplete, as defined in the header file.")
+
+
+def print_unimplemented_functions(dif_statuses):
+    """Print a table of unimplemented DIF functions to STDOUT."""
+    rows = []
+    headers = ["IP", "Function"]
+    for dif_status in dif_statuses:
+        if not dif_status.api_complete:
+            rows.append(
+                [dif_status.ip, "\n".join(dif_status.funcs_unimplemented)])
+    print("Unimplemented Functions:")
+    print(tabulate(rows, headers, tablefmt=TABLE_FORMAT))
+
+
+def main(argv):
+    # Process args.
+    parser = argparse.ArgumentParser(
+        prog="check_dif_statuses",
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument(
+        "repo_top",
+        help=
+        """Relative path to where the OpenTitan repository is checked out.""")
+    parser.add_argument(
+        "--show-unimplemented",
+        action='store_true',
+        help="""Show unimplemented functions for each incomplete DIF.""")
+    args = parser.parse_args(argv)
+
+    # Define root path of DIFs.
+    difs_root_path = os.path.join("sw", "device", "lib", "dif")
+
+    # Get list of all DIF basenames.
+    shared_headers = ["dif_warn_unused_result"]
+    difs = get_list_of_difs(difs_root_path, shared_headers)
+
+    # Get DIF statuses (while displaying progress bar).
+    dif_statuses = []
+    print("Analyzing statuses of %s DIFs ..." % len(difs))
+    progress_bar_width = len(difs)
+    sys.stdout.write("[%s]" % (" " * progress_bar_width))
+    sys.stdout.flush()
+    sys.stdout.write(
+        "\b" * (progress_bar_width + 1))  # return to start of line, after '['
+    for dif in difs:
+        # Process a DIF.
+        dif_statuses.append(DIFStatus(args.repo_top, difs_root_path, dif))
+        # Update the progress bar.
+        sys.stdout.write("#")
+        sys.stdout.flush()
+    sys.stdout.write("]\n")  # end the progress bar
```

Comment:
I originally chose a manual progress bar instead of tqdm b/c I didn't want to bloat the python-requirements.txt file but it seems tqdm is already a dependency of another packages so no sense not to use it. Nice catch!